File: sdk/go/doctor.go
// Package llmclient - Doctor health check
// 用法：
//
//	result := c.Doctor(ctx)
//	result.Print()
//	if !result.OK { log.Fatal("LLM SDK 配置异常") }
package llmclient

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

// DoctorCheck is the result of a single health check.
type DoctorCheck struct {
	Name    string
	OK      bool
	Message string
	Fix     string // remediation hint, empty when OK=true
}

// DoctorResult holds all check results from Doctor().
type DoctorResult struct {
	Checks []DoctorCheck
}

// OK returns true if every check passed.
func (r *DoctorResult) OK() bool {
	for _, c := range r.Checks {
		if !c.OK {
			return false
		}
	}
	return true
}

// Print writes a human-readable diagnostic report to stdout.
func (r *DoctorResult) Print() {
	fmt.Println("LLM SDK Doctor")
	fmt.Println(strings.Repeat("=", 40))
	for _, c := range r.Checks {
		icon := "✓"
		if !c.OK {
			icon = "✗"
		}
		fmt.Printf("%s  %-22s %s\n", icon, c.Name, c.Message)
		if !c.OK && c.Fix != "" {
			fmt.Printf("   %-22s → %s\n", "", c.Fix)
		}
	}
	fmt.Println()
	if r.OK() {
		fmt.Println("全部通过，SDK 可正常使用。")
	} else {
		failed := 0
		for _, c := range r.Checks {
			if !c.OK {
				failed++
			}
		}
		fmt.Printf("%d 项检查未通过，请按上方提示修复。\n", failed)
	}
}

// Doctor runs all health checks and returns a DoctorResult.
// It is safe to call concurrently. Each check has a short timeout
// so Doctor returns quickly even when the Proxy is unreachable.
func (c *Client) Doctor(ctx context.Context) *DoctorResult {
	result := &DoctorResult{}
	httpClient := &http.Client{Timeout: 5 * time.Second}

	// ── 1. LLM_BASE_URL 是否配置 ─────────────────────────────────────────────
	envURL := os.Getenv("LLM_BASE_URL")
	if envURL != "" {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "LLM_BASE_URL",
			OK:      true,
			Message: fmt.Sprintf("已配置 (%s)", c.BaseURL),
		})
	} else {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "LLM_BASE_URL",
			OK:      false,
			Message: fmt.Sprintf("未设置，使用默认值 (%s)", c.BaseURL),
			Fix:     "export LLM_BASE_URL=https://your-proxy.example.com  （或启动本地 Proxy）",
		})
	}

	// ── 2. LLM_API_KEY 是否配置 ──────────────────────────────────────────────
	envKey := os.Getenv("LLM_API_KEY")
	if envKey != "" && envKey != "no-key" {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "LLM_API_KEY",
			OK:      true,
			Message: "已配置",
		})
	} else {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "LLM_API_KEY",
			OK:      false,
			Message: "未设置或使用占位值 'no-key'",
			Fix:     "export LLM_API_KEY=sk-your-team-key  （向平台团队申请）",
		})
	}

	// ── 3. Proxy 可达性检查 ───────────────────────────────────────────────────
	proxyReachable := false
	readinessURL := c.BaseURL + "/health/readiness"
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, readinessURL, nil)
	resp, err := httpClient.Do(req)
	if err != nil {
		fix := "共享 Proxy：确认 LLM_BASE_URL 正确，检查网络/VPN\n   本地 Proxy：docker compose -f proxy/docker-compose.dev.yaml up -d"
		if strings.Contains(err.Error(), "timeout") || strings.Contains(err.Error(), "context deadline") {
			fix = "检查网络连接，确认 Proxy 地址可从本机访问"
		}
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "Proxy 可达",
			OK:      false,
			Message: fmt.Sprintf("连接失败 (%s): %v", c.BaseURL, err),
			Fix:     fix,
		})
	} else {
		resp.Body.Close()
		if resp.StatusCode < 500 {
			proxyReachable = true
			result.Checks = append(result.Checks, DoctorCheck{
				Name:    "Proxy 可达",
				OK:      true,
				Message: fmt.Sprintf("HTTP %d /health/readiness", resp.StatusCode),
			})
		} else {
			result.Checks = append(result.Checks, DoctorCheck{
				Name:    "Proxy 可达",
				OK:      false,
				Message: fmt.Sprintf("HTTP %d，Proxy 返回错误", resp.StatusCode),
				Fix:     "检查 Proxy 服务是否正常运行，查看 Proxy 日志",
			})
		}
	}

	// ── 4. 鉴权有效性 ─────────────────────────────────────────────────────────
	// 用 GET /models 验证 key，避免触发实际推理（不依赖具体模型是否可用）
	if proxyReachable {
		authReq, _ := http.NewRequestWithContext(ctx, http.MethodGet,
			c.BaseURL+"/models", nil)
		authReq.Header.Set("Authorization", "Bearer "+c.APIKey)
		authClient := &http.Client{Timeout: 10 * time.Second}
		authResp, err := authClient.Do(authReq)
		if err != nil {
			result.Checks = append(result.Checks, DoctorCheck{
				Name:    "鉴权有效",
				OK:      false,
				Message: fmt.Sprintf("请求失败: %v", err),
				Fix:     "确认 Proxy 可用后重试",
			})
		} else {
			io.Copy(io.Discard, authResp.Body)
			authResp.Body.Close()
			switch authResp.StatusCode {
			case 200, 201:
				result.Checks = append(result.Checks, DoctorCheck{
					Name:    "鉴权有效",
					OK:      true,
					Message: "API key 验证成功",
				})
			case 401, 403:
				result.Checks = append(result.Checks, DoctorCheck{
					Name:    "鉴权有效",
					OK:      false,
					Message: fmt.Sprintf("HTTP %d 鉴权失败", authResp.StatusCode),
					Fix:     "检查 LLM_API_KEY 是否与 Proxy 端 LITELLM_MASTER_KEY 一致，向平台团队确认",
				})
			default:
				result.Checks = append(result.Checks, DoctorCheck{
					Name:    "鉴权有效",
					OK:      false,
					Message: fmt.Sprintf("HTTP %d", authResp.StatusCode),
					Fix:     "查看 Proxy 日志了解详情",
				})
			}
		}
	} else {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "鉴权有效",
			OK:      false,
			Message: "跳过（Proxy 不可达）",
			Fix:     "先修复 Proxy 可达性问题",
		})
	}

	// ── 5. Capability override 合理性检查 ────────────────────────────────────
	// 含 / 是合法 provider/model 格式；含 - 是 Proxy 内部别名，两者都允许。
	// 没有 / 也没有 - 的短名才可能是错误格式。
	tagEnvMap := map[string]string{
		"chat":        "LLM_MODEL_CHAT",
		"vision":      "LLM_MODEL_VISION",
		"video-input": "LLM_MODEL_VIDEO",
		"embedding":   "LLM_MODEL_EMBEDDING",
		"image-gen":   "LLM_MODEL_IMAGE_GEN",
		"fast":        "LLM_MODEL_FAST",
		"local":       "LLM_MODEL_LOCAL",
	}
	overrideIssues := 0
	for tag, envKeyName := range tagEnvMap {
		_ = tag
		envVal := os.Getenv(envKeyName)
		if envVal != "" && !strings.Contains(envVal, "/") && !strings.Contains(envVal, "-") {
			overrideIssues++
		}
	}

	if overrideIssues > 0 {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "Capability 配置",
			OK:      false,
			Message: fmt.Sprintf("发现 %d 个可能有问题的覆盖", overrideIssues),
			Fix:     "模型名建议使用 provider/model 格式，如 openai/gpt-4o 或使用默认值",
		})
	} else {
		result.Checks = append(result.Checks, DoctorCheck{
			Name:    "Capability 配置",
			OK:      true,
			Message: fmt.Sprintf("%d 个 tag 配置正常", len(c.TagMap)),
		})
	}

	return result
}

File: sdk/typescript/doctor.ts
/**
 * LLM SDK Doctor - TypeScript
 * ============================
 * 连接与配置健康检查
 *
 * 用法:
 *   import { client } from '@goat/llm-sdk'
 *   const result = await client.doctor()
 *   result.print()
 *
 *   // 或断言全部通过
 *   if (!result.ok) throw new Error('LLM SDK 配置异常')
 */

/** 单项检查结果 */
export interface DoctorCheck {
  name: string
  ok: boolean
  message: string
  fix?: string  // 修复建议，仅在 ok=false 时填充
}

/** doctor() 的完整诊断结果 */
export class DoctorResult {
  constructor(public readonly checks: DoctorCheck[] = []) {}

  get ok(): boolean {
    return this.checks.every(c => c.ok)
  }

  print(): void {
    console.log('LLM SDK Doctor')
    console.log('='.repeat(40))
    for (const c of this.checks) {
      const icon = c.ok ? '✓' : '✗'
      const namePad = c.name.padEnd(22)
      console.log(`${icon}  ${namePad} ${c.message}`)
      if (!c.ok && c.fix) {
        console.log(`   ${''.padEnd(22)} → ${c.fix}`)
      }
    }
    console.log()
    if (this.ok) {
      console.log('全部通过，SDK 可正常使用。')
    } else {
      const failed = this.checks.filter(c => !c.ok).length
      console.log(`${failed} 项检查未通过，请按上方提示修复。`)
    }
  }
}

const tagEnvMap: Record<string, string> = {
  'chat':        'LLM_MODEL_CHAT',
  'vision':      'LLM_MODEL_VISION',
  'video-input': 'LLM_MODEL_VIDEO',
  'embedding':   'LLM_MODEL_EMBEDDING',
  'image-gen':   'LLM_MODEL_IMAGE_GEN',
  'fast':        'LLM_MODEL_FAST',
  'local':       'LLM_MODEL_LOCAL',
}

/**
 * 执行所有健康检查，返回 DoctorResult。
 * 设计原则：每项检查独立，超时设短（5s），错误给出 fix hint。
 */
export async function runDoctor(
  baseURL: string,
  apiKey: string,
  tagMap: Record<string, string>,
): Promise<DoctorResult> {
  const checks: DoctorCheck[] = []

  // ── 1. LLM_BASE_URL 是否配置 ──────────────────────────────────────────────
  const envURL = process.env.LLM_BASE_URL ?? ''
  if (envURL) {
    checks.push({ name: 'LLM_BASE_URL', ok: true, message: `已配置 (${baseURL})` })
  } else {
    checks.push({
      name: 'LLM_BASE_URL',
      ok: false,
      message: `未设置，使用默认值 (${baseURL})`,
      fix: 'export LLM_BASE_URL=https://your-proxy.example.com  （或启动本地 Proxy）',
    })
  }

  // ── 2. LLM_API_KEY 是否配置 ───────────────────────────────────────────────
  const envKey = process.env.LLM_API_KEY ?? ''
  if (envKey && envKey !== 'no-key') {
    checks.push({ name: 'LLM_API_KEY', ok: true, message: '已配置' })
  } else {
    checks.push({
      name: 'LLM_API_KEY',
      ok: false,
      message: "未设置或使用占位值 'no-key'",
      fix: 'export LLM_API_KEY=sk-your-team-key  （向平台团队申请）',
    })
  }

  // ── 3. Proxy 可达性检查 ───────────────────────────────────────────────────
  let proxyReachable = false
  try {
    const controller = new AbortController()
    const timeout = setTimeout(() => controller.abort(), 5000)
    const resp = await fetch(`${baseURL}/health/readiness`, {
      signal: controller.signal,
    }).finally(() => clearTimeout(timeout))

    if (resp.status < 500) {
      proxyReachable = true
      checks.push({
        name: 'Proxy 可达',
        ok: true,
        message: `HTTP ${resp.status} /health/readiness`,
      })
    } else {
      checks.push({
        name: 'Proxy 可达',
        ok: false,
        message: `HTTP ${resp.status}，Proxy 返回错误`,
        fix: '检查 Proxy 服务是否正常运行，查看 Proxy 日志',
      })
    }
  } catch (e: any) {
    const isTimeout = e?.name === 'AbortError'
    if (isTimeout) {
      checks.push({
        name: 'Proxy 可达',
        ok: false,
        message: `连接超时 (${baseURL})`,
        fix: '检查网络连接，确认 Proxy 地址可从本机访问',
      })
    } else {
      checks.push({
        name: 'Proxy 可达',
        ok: false,
        message: `连接失败 (${baseURL}): ${e?.message ?? e}`,
        fix: '共享 Proxy：确认 LLM_BASE_URL 正确，检查网络/VPN\n   本地 Proxy：docker compose -f proxy/docker-compose.dev.yaml up -d',
      })
    }
  }

  // ── 4. 鉴权有效性 ─────────────────────────────────────────────────────────
  // 用 GET /models 验证 key，避免触发实际推理（不依赖具体模型是否可用）
  if (proxyReachable) {
    try {
      const controller = new AbortController()
      const timeout = setTimeout(() => controller.abort(), 10000)
      const resp = await fetch(`${baseURL}/models`, {
        headers: { Authorization: `Bearer ${apiKey}` },
        signal: controller.signal,
      }).finally(() => clearTimeout(timeout))

      if (resp.status === 200 || resp.status === 201) {
        checks.push({ name: '鉴权有效', ok: true, message: 'API key 验证成功' })
      } else if (resp.status === 401 || resp.status === 403) {
        checks.push({
          name: '鉴权有效',
          ok: false,
          message: `HTTP ${resp.status} 鉴权失败`,
          fix: '检查 LLM_API_KEY 是否与 Proxy 端 LITELLM_MASTER_KEY 一致，向平台团队确认',
        })
      } else {
        checks.push({
          name: '鉴权有效',
          ok: false,
          message: `HTTP ${resp.status}`,
          fix: '查看 Proxy 日志了解详情',
        })
      }
    } catch (e: any) {
      checks.push({
        name: '鉴权有效',
        ok: false,
        message: `请求失败: ${e?.message ?? e}`,
        fix: '确认 Proxy 可用后重试',
      })
    }
  } else {
    checks.push({
      name: '鉴权有效',
      ok: false,
      message: '跳过（Proxy 不可达）',
      fix: '先修复 Proxy 可达性问题',
    })
  }

  // ── 5. Capability override 合理性检查 ────────────────────────────────────
  // 含 / 是合法 provider/model 格式；含 - 是 Proxy 内部别名，两者都允许。
  // 没有 / 也没有 - 的短名才可能是错误格式。
  const overrideIssues: string[] = []
  for (const [tag, model] of Object.entries(tagMap)) {
    const envKeyName = tagEnvMap[tag]
    const envVal = envKeyName ? (process.env[envKeyName] ?? '') : ''
    if (envVal && !envVal.includes('/') && !envVal.includes('-')) {
      overrideIssues.push(`${envKeyName}=${envVal} 可能不含 provider 前缀`)
    }
  }

  if (overrideIssues.length > 0) {
    checks.push({
      name: 'Capability 配置',
      ok: false,
      message: `发现 ${overrideIssues.length} 个可能有问题的覆盖`,
      fix: '模型名建议使用 provider/model 格式，如 openai/gpt-4o 或使用默认值',
    })
  } else {
    checks.push({
      name: 'Capability 配置',
      ok: true,
      message: `${Object.keys(tagMap).length} 个 tag 配置正常`,
    })
  }

  return new DoctorResult(checks)
}
